File: parte2_gestor_academico/gestor_academico_base.cpp
// Código monolítico defectuoso en C++ (Base para la Parte 2)

#include <iostream>
#include <vector>
#include <string>
#include <fstream>

struct EstudianteBase {
    std::string codigo;
    std::string nombre;
    std::vector<double> notas;
};

class GestorAcademicoMonolitico {
public:
    std::vector<EstudianteBase> estudiantes; // Estado expuesto

    void agregarEstudiante(const std::string& codigo, const std::string& nombre) {
        estudiantes.push_back({codigo, nombre, {}});
        std::ofstream auditoria("auditoria.txt", std::ios::app);
        auditoria << "Estudiante agregado: " << codigo << " - " << nombre << "\n";
    }

    void agregarNota(const std::string& codigo, double nota) {
        // No valida que la nota esté entre 0 y 20
        for (auto& est : estudiantes) {
            if (est.codigo == codigo) {
                est.notas.push_back(nota);
                std::ofstream auditoria("auditoria.txt", std::ios::app);
                auditoria << "Nota " << nota << " agregada a " << codigo << "\n";
            }
        }
    }

    void generarReporte() {
        std::ofstream reporte("reporte.csv");
        reporte << "codigo,nombre,promedio,estado\n";
        for (const auto& est : estudiantes) {
            double suma = 0.0;
            for (double n : est.notas) suma += n;
            double promedio = est.notas.empty() ? 0.0 : suma / est.notas.size();
            std::string estado = (promedio >= 10.5) ? "APROBADO" : "DESAPROBADO";
            reporte << est.codigo << "," << est.nombre << "," << promedio << "," << estado << "\n";
        }
    }
};

File: parte2_gestor_academico/gestor_academico_base.py
# Código monolítico defectuoso (Base para la Parte 2)

class GestorAcademicoMonolitico:
    def __init__(self):
        self.estudiantes = []  # Lista pública (Estado expuesto)

    def agregar_estudiante(self, codigo, nombre):
        self.estudiantes.append({"codigo": codigo, "nombre": nombre, "notas": []})
        # Auditoría acoplada directamente a un archivo de texto
        with open("auditoria.txt", "a", encoding="utf-8") as f:
            f.write(f"Estudiante agregado: {codigo} - {nombre}\n")

    def agregar_nota(self, codigo, nota):
        # Sin validación de rango de nota ([0, 20])
        for est in self.estudiantes:
            if est["codigo"] == codigo:
                est["notas"].append(nota)
                with open("auditoria.txt", "a", encoding="utf-8") as f:
                    f.write(f"Nota {nota} agregada a {codigo}\n")

    def generar_reporte(self):
        # Múltiples responsabilidades: cálculo + generación de archivo I/O
        with open("reporte.csv", "w", encoding="utf-8") as f:
            f.write("codigo,nombre,promedio,estado\n")
            for est in self.estudiantes:
                notas = est["notas"]
                promedio = sum(notas) / len(notas) if notas else 0.0
                estado = "APROBADO" if promedio >= 10.5 else "DESAPROBADO"
                f.write(f"{est['codigo']},{est['nombre']},{promedio:.2f},{estado}\n")